app/code/Magento/Checkout/Model/TotalsInformationManagement.php

```diff
@@ -57,6 +57,7 @@ public function calculate(
             );
         }
         $quote->collectTotals();
+        $this->cartRepository->save($quote);
```

Comment:
When I was thinking about the fix I had a look at the workflow of "Update Shopping Cart" functionality. It calls ```\Magento\Checkout\Model\Cart::save()```
There after collectTotals() method executes ```$this->quoteRepository->save($this->getQuote());```

Reply:
Hi @paliarush,

I'm not sure if it's ok to have cart totals save instead of just collect totals

Could you review and confirm this behavior change?

Reply:
Hi @zhartaunik,

It looks like TotalsInformationManagement model in method "calculate" shouldn't save quote, it should execute only calculations. 

Use case: I'd like to _calculate_ how many will I pay for shipment if I'll change region from New York to California --> I'll execute this method, but it doesn't mean I want to save it.

Calculation and re-saving should be still different actions, otherwise it might get performance issue or even worth - in some case infinite recursion.

Could you describe why the way of adding `trigger_recollect` flag was bad and why it didn't worked fine?

Reply:
This method definitely should not save the quote. Probably the best solution is to try to fix it while collecting the totals. @zhartaunik Could dig int that direction?